### PR TITLE
Add support for multiple public certificates to AWS IID node attestor

### LINKED ISSUE #2126: AWS node attestor needs to use different cert for different regions
* **Version**: 3ee20c95bbf2e96436646f56ef1e3f8508a61a37
* **Platform**: GCP
* **Subsystem**: node attestor plugin for aws

We recently spun up an AWS vm in eu-south-1 (Milan), our existing spire-server (running from GCP) failed to attest the new VM with "cryptography signature verification failures".

It seems that the aws_iid plugin is using a hardcoded [cert](https://github.com/spiffe/spire/blob/master/pkg/server/plugin/nodeattestor/aws/iid.go#L59) regardless of the region the aws vm lives in.

However, according to aws [doc](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/verify-pkcs7.html) (for PKCS7 signature verification), milan region has a different DSA cert. I suspect the should have a different RSA cert for that region as well (which I believe is what aws_iid code is using).

It seems not correct to hard code the cert to be used for all region's node attestation.

### PR DESCRIPTION
**Pull request by maxlambrecht**
**Pull Request check list**

- [X] Commit conforms to CONTRIBUTING.md?
- [X] Proper tests/regressions included?
- [ ] Documentation updated?

**Affected functionality**

Add support for all regions to the `AWS_IID` Node Attestor.

**Description of change**

- Add AWS public RSA-2048 certificates for all regions.
- Select the public certificate using the region code from the Attestation Data.
- Use RSA-2048 certificates to verify PKCS#7 signatures: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/verify-rsa2048.html

**Backward compatibility** 
 
The SPIRE Server also supports the use of RSA-1024 certificates: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/verify-signature.html
The "default" certificate is the one currently used by SPIRE, which expires on Jun-2024. The certificates for specific regions were also added as part of this PR.

SPIRE Agent will send both type of signatures: the signature from the path `instance-identity/signature`, that can be validated using RSA-1024 certificates); and the the signature from the path `instance-identity/rsa2048`, that can be validated using RSA-2048 certificates). 

SPIRE Server will use the RSA-2048 signature if present, if not, will fall back to RSA-1024. 

_**In a nutshell:**_
*New* SPIRE Server attests *new* SPIRE Agents, using RSA-2048 certificates and signature.  
*Old* SPIRE Server attests *new* SPIRE Agents, using only one "default" RSA-1024 certificate, supporting only a set of regions. 
*New* SPIRE Server attests *old* SPIRE Agents, using RSA-1024 certificates, supporting all regions. 

**Which issue this PR fixes**

Fixes #2126

### CODE REVIEW

File: pkg/server/plugin/nodeattestor/awsiid/awsca.go [hunk 1]
```diff
@@ -1,42 +1,48 @@
...
+	RSA1024
+	RSA2048
 )
 
-func getAWSCAPublicKey() (*rsa.PublicKey, error) {
-	ca, err := pemutil.ParseCertificate([]byte(awsCACertPEM))
+func getAWSCACertificate(region string, keyType PublicKeyType) (*x509.Certificate, error) {
+	var cert string
+	if keyType == KeyTypeUnset {
+		return nil, fmt.Errorf("signature type is unset")
```
Comment (azdagron) [inline pkg/server/plugin/nodeattestor/awsiid/awsca.go:24]:
```suggestion
		return nil, fmt.Errorf("signature key type is unset")
```

File: pkg/server/plugin/nodeattestor/awsiid/iid.go [hunk 2]
```diff
@@ -223,14 +227,10 @@ func (p *IIDAttestorPlugin) Configure(ctx context.Context, req *configv1.Configu
 		return nil, status.Errorf(codes.InvalidArgument, "unable to decode configuration: %v", err)
 	}
 
-	// Get the AWS CA public key. We do this lazily on configure so deployments
+	// Function to get the AWS CA certificate. We do this lazily on configure so deployments
```
Comment (azdagron) [inline pkg/server/plugin/nodeattestor/awsiid/iid.go:230]: Did you measure the startup cost if we parse all of the certificates in advance? I'd be surprised if the impact was substantial. I'd also much prefer to pay that cost once instead of every time we attest a new node. My vote would be to either 1) just parse them all on startup, or 2) cache them once parsed so we don't pay the cost continuously.

File: pkg/server/plugin/nodeattestor/awsiid/iid.go [hunk 3]
```diff
@@ -106,14 +110,14 @@ type IIDAttestorConfig struct {
 	AssumeRole                      string   `hcl:"assume_role"`
 	pathTemplate                    *agentpathtemplate.Template
 	trustDomain                     spiffeid.TrustDomain
-	awsCAPublicKey                  *rsa.PublicKey
+	getAWSCACertificate             func(string, PublicKeyType) (*x509.Certificate, error)
```
Comment (azdagron) [inline pkg/server/plugin/nodeattestor/awsiid/iid.go:113]: We talked about maybe providing a way in the config for operators to specify an override for the signing certificate for a given region. Is that not needed anymore due to the very long expiration dates on the RSA2048 CAs?

File: pkg/server/plugin/nodeattestor/awsiid/iid.go [hunk 4]
```diff
@@ -335,31 +337,100 @@ func tagsFromInstance(instance ec2types.Instance) instanceTags {
 	return tags
 }
 
-func unmarshalAndValidateIdentityDocument(data []byte, pubKey *rsa.PublicKey) (imds.InstanceIdentityDocument, error) {
+func unmarshalAndValidateIdentityDocument(data []byte, getAWSCACertificate func(string, PublicKeyType) (*x509.Certificate, error)) (imds.InstanceIdentityDocument, error) {
 	var attestationData caws.IIDAttestationData
+	var doc imds.InstanceIdentityDocument
```
Comment (azdagron) [inline pkg/server/plugin/nodeattestor/awsiid/iid.go:342]: nitpick: variables should be defined closer to the point of use

File: pkg/server/plugin/nodeattestor/awsiid/iid_test.go [hunk 5]
```diff
@@ -56,8 +60,14 @@ var (
 	ebsType              = ec2types.DeviceTypeEbs
 )
 
+var testAWSCACert *x509.Certificate
```
Comment (azdagron) [inline pkg/server/plugin/nodeattestor/awsiid/iid_test.go:63]: nitpick: these two vars should go in the var block above

**Review by azdagron — COMMENTED** [review]
Thanks, @maxlambrecht ! Just a few comments...

File: pkg/server/plugin/nodeattestor/awsiid/iid.go (hunk 3 above)
Comment (maxlambrecht) [reply pkg/server/plugin/nodeattestor/awsiid/iid.go:113]: We talked about providing a way to specify a cert for region, and my understanding is that we prefer to have all the certs and select automatically which one to use based on the region, which I think is not bad idea considering the very long expiration dates.

File: pkg/server/plugin/nodeattestor/awsiid/iid.go (hunk 2 above)
Comment (maxlambrecht) [reply pkg/server/plugin/nodeattestor/awsiid/iid.go:230]: I prefer option 2. Rather than parsing all 45 certificates on startup, I'd choose parsing and caching the certificates on demand. This way, only the certificates that are actually used will be parsed and cached, avoiding unnecessary overhead.

**Review by azdagron — APPROVED** [review]
Thanks @maxlambrecht !
